spark/src/main/scala/com/gu/tableversions/spark/filesystem/PathMapper.scala

```diff
@@ -0,0 +1,91 @@
+package com.gu.tableversions.spark.filesystem
+
+import java.net.URI
+
+import com.gu.tableversions.core.{Partition, Version}
+import org.apache.hadoop.fs.Path
+
+/**
+  * Map paths between one "outer" FileSystem (e.g. a ProxyFileSystem) and an underlying one.
+  * For example, by changing the schema, or appending version directories.
+  */
+trait PathMapper {
+
+  /** Convert a path with a "versioned" scheme to one suitable for the underlying FileSystem. */
+  def forUnderlying(path: Path): Path
+
+  /** Convert a path from the underlying FileSystem to one in the "versioned" scheme. */
+  def fromUnderlying(path: Path): Path
+}
+
+/**
+  * Path mapper that converts paths for partitions according to a pre-configured set of versioned partitions.
+  */
+class VersionedPathMapper(underlyingFsScheme: String, partitionVersions: Map[Partition, Version]) extends PathMapper {
+
+  import VersionedPathMapper._
+
+  override def forUnderlying(path: Path): Path = {
+    assert(path.toUri.getScheme == VersionedFileSystem.scheme,
+           s"Path provided to `forUnderlying` ($path) not in the ${VersionedFileSystem.scheme} scheme")
```

Comment:
Can we avoid runtime check in favor compile time? aka `Path` type is used to represents two different things. Can they be two different type so one cannot construct the wrong type of Path?

Reply:
I'm not sure we can in any useful way. The path mapper converts the paths that are given to it by the `ProxyFilesystem`. If we made this API more strongly typed, the filesystem would have to wrap the `Path`s it gets from the calling code (in Spark) to the right type. But I don't think this would be less error prone than making sure to call the right method in the first place.